python/mirascope/llm/prompts/prompts.py

```diff
@@ -0,0 +1,424 @@
+        """Generates a response using the provided model."""
+        return self.call(model, *args, **kwargs)  # pragma: no cover
+
+    @overload
+    def call(
+        self: "Prompt[P, None]", model: Model, *args: P.args, **kwargs: P.kwargs
+    ) -> Response: ...
+
+    @overload
+    def call(
+        self: "Prompt[P, FormattableT]", model: Model, *args: P.args, **kwargs: P.kwargs
+    ) -> Response[FormattableT]: ...
+
+    def call(
+        self, model: Model, *args: P.args, **kwargs: P.kwargs
+    ) -> Response | Response[FormattableT]:
+        """Generates a response using the provided model."""
+        result = self.fn(*args, **kwargs)
+        messages = _utils.promote_to_messages(result)
+        return model.call(messages=messages, tools=self.toolkit, format=self.format)
+
+    @overload
+    def stream(
+        self: "Prompt[P, None]", model: Model, *args: P.args, **kwargs: P.kwargs
+    ) -> StreamResponse: ...
+
+    @overload
+    def stream(
+        self: "Prompt[P, FormattableT]", model: Model, *args: P.args, **kwargs: P.kwargs
+    ) -> StreamResponse[FormattableT]: ...
+
+    def stream(
+        self, model: Model, *args: P.args, **kwargs: P.kwargs
+    ) -> StreamResponse | StreamResponse[FormattableT]:
+        """Generates a streaming response using the provided model."""
+        result = self.fn(*args, **kwargs)
+        messages = _utils.promote_to_messages(result)
+        return model.stream(messages=messages, tools=self.toolkit, format=self.format)
+
+
+@dataclass
+class AsyncPrompt(Generic[P, FormattableT]):
+    """An async prompt that can be called with a model to generate a response.
+
+    Created by decorating an async `PromptFunction` with `llm.prompt`. The decorated
+    async function becomes callable with a `Model` to generate LLM responses asynchronously.
+
+    An `AsyncPrompt` is essentially: async `PromptFunction` + tools + format.
+    It can be invoked with a model: `await prompt(model, *args, **kwargs)`.
+    """
+
+    fn: AsyncPromptFunction[P]
+    """The underlying async prompt function that generates message content."""
+
+    toolkit: AsyncToolkit
+    """The toolkit containing this prompt's async tools."""
+
+    format: type[FormattableT] | Format[FormattableT] | None
+    """The response format for the generated response."""
+
+    @overload
+    async def __call__(
+        self: "AsyncPrompt[P, None]", model: Model, *args: P.args, **kwargs: P.kwargs
+    ) -> AsyncResponse: ...
+
+    @overload
+    async def __call__(
+        self: "AsyncPrompt[P, FormattableT]",
+        model: Model,
+        *args: P.args,
+        **kwargs: P.kwargs,
+    ) -> AsyncResponse[FormattableT]: ...
+
+    async def __call__(
+        self, model: Model, *args: P.args, **kwargs: P.kwargs
+    ) -> AsyncResponse | AsyncResponse[FormattableT]:
+        """Generates a response using the provided model asynchronously."""
+        return await self.call(model, *args, **kwargs)  # pragma: no cover
```

Comment:
I didn't see a need to have unit test coverage for the \__call__ redirect, hence the no covers.